### PR TITLE
Add new view-webhook plugin

### PR DESCRIPTION
**Pull request by developer-guy**
### Submitting a new plugin
[x] Make sure you read the Plugin Naming Guide: https://krew.sigs.k8s.io/docs/developer-guide/develop/naming-guide/
[x] Verify you can install your plugin locally: kubectl krew install --manifest=[...] --archive=[...]

### Description

* This project aims to visualize critical parts of the admission webhook configuration resource.

See Github Repo: https://github.com/Trendyol/kubectl-view-webhook
Screen Shot 2020-11-12 at 14 49 59

### CODE REVIEW

File: plugins/view-webhook.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+apiVersion: krew.googlecontainertools.github.com/v1alpha2
+kind: Plugin
+metadata:
+  name: view-webhook
+spec:
+  version: v0.0.3
+  homepage: https://github.com/Trendyol/kubectl-view-webhook
+  shortDescription: Visualize your webhook configurations in Kubernetes.
+  description: |
+    👀 This projects aims to visualize critical parts of the admission webhook configuration resource
```
Comment (ahmetb) [inline plugins/view-webhook.yaml:10]: - Let's avoid emojis. Many terminals still can't render them.
- Please line wrap around 80 chars.
- `This projects aims to` doesn't add anything, remove.

File: plugins/view-webhook.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+apiVersion: krew.googlecontainertools.github.com/v1alpha2
+kind: Plugin
+metadata:
+  name: view-webhook
+spec:
+  version: v0.0.3
+  homepage: https://github.com/Trendyol/kubectl-view-webhook
+  shortDescription: Visualize your webhook configurations in Kubernetes.
```
Comment (ahmetb) [inline plugins/view-webhook.yaml:8]: `Visualize your webhook configurations` is sufficient. It's clear kubectl plugins are for Kubernetes. :)

**Comment by ahmetb** [comment]
@developer-guy thanks for submission, it is a fit for Krew-index.

However, I can't use this plugin:

> Error: no Auth Provider found for name "gcp"

This is explained at https://krew.sigs.k8s.io/docs/developer-guide/develop/best-practices/#auth-plugins

We can merge and fix this later, but it is preferred if you can fix now.
/hold

File: plugins/view-webhook.yaml (hunk 1 above)
Comment (developer-guy) [reply plugins/view-webhook.yaml]: Thanks to your valuable comments, done btw. 🙏

File: plugins/view-webhook.yaml (hunk 2 above)
Comment (developer-guy) [reply plugins/view-webhook.yaml]: Thanks to your valuable comments, done btw. 🙏

**Comment by developer-guy** [comment]
> @developer-guy thanks for submission, it is a fit for Krew-index.
> 
> However, I can't use this plugin:
> 
> > Error: no Auth Provider found for name "gcp"
> 
> This is explained at [krew.sigs.k8s.io/docs/developer-guide/develop/best-practices/#auth-plugins](https://krew.sigs.k8s.io/docs/developer-guide/develop/best-practices/#auth-plugins)
> 
> We can merge and fix this later, but it is preferred if you can fix now.
> /hold

I already fixed at the v0.0.4 release of the project, thank you. 👍

**Comment by ahmetb** [comment]
I now got this with 0.0.4:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1e94153]

goroutine 1 [running]:
github.com/Trendyol/kubectl-view-webhook/pkg/k8s.(*WebHookClient).fillMutatingWebhookConfigurations(0xc0000b6400, 0x0, 0x0, 0x0, 0x0, 0xc0002828c0, 0x32, 0x0, 0x0, 0x0, ...)
	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/pkg/k8s/mutating_webhook.go:92 +0x1f3
github.com/Trendyol/kubectl-view-webhook/pkg/k8s.(*WebHookClient).Run(0xc0000b6400, 0xc00003a1d0, 0x1, 0x1, 0xd, 0x2236315, 0x5)
	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/pkg/k8s/mutating_webhook.go:61 +0x208
github.com/Trendyol/kubectl-view-webhook/cmd.(*ViewWebhookOptions).Run(0xc0003d96c0, 0xc0003bd340, 0x2cf7778)
	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/cmd/root.go:133 +0xa5
github.com/Trendyol/kubectl-view-webhook/cmd.NewCmdViewWebhook.func1(0xc0003bd340, 0x2cf7778, 0x0, 0x0, 0x0, 0x0)
	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/cmd/root.go:70 +0x78
github.com/spf13/cobra.(*Command).execute(0xc0003bd340, 0xc00003a1d0, 0x0, 0x0, 0xc0003bd340, 0xc00003a1d0)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0xc0003bd340, 0xc00000e010, 0x23d4ce0, 0xc00000e018)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
main.main()
	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/main.go:32 +0x105
```

**Comment by developer-guy** [comment]
> I now got this with 0.0.4:
> 
> ```
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1e94153]
> 
> goroutine 1 [running]:
> github.com/Trendyol/kubectl-view-webhook/pkg/k8s.(*WebHookClient).fillMutatingWebhookConfigurations(0xc0000b6400, 0x0, 0x0, 0x0, 0x0, 0xc0002828c0, 0x32, 0x0, 0x0, 0x0, ...)
> 	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/pkg/k8s/mutating_webhook.go:92 +0x1f3
> github.com/Trendyol/kubectl-view-webhook/pkg/k8s.(*WebHookClient).Run(0xc0000b6400, 0xc00003a1d0, 0x1, 0x1, 0xd, 0x2236315, 0x5)
> 	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/pkg/k8s/mutating_webhook.go:61 +0x208
> github.com/Trendyol/kubectl-view-webhook/cmd.(*ViewWebhookOptions).Run(0xc0003d96c0, 0xc0003bd340, 0x2cf7778)
> 	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/cmd/root.go:133 +0xa5
> github.com/Trendyol/kubectl-view-webhook/cmd.NewCmdViewWebhook.func1(0xc0003bd340, 0x2cf7778, 0x0, 0x0, 0x0, 0x0)
> 	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/cmd/root.go:70 +0x78
> github.com/spf13/cobra.(*Command).execute(0xc0003bd340, 0xc00003a1d0, 0x0, 0x0, 0xc0003bd340, 0xc00003a1d0)
> 	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x47c
> github.com/spf13/cobra.(*Command).ExecuteC(0xc0003bd340, 0xc00000e010, 0x23d4ce0, 0xc00000e018)
> 	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x375
> github.com/spf13/cobra.(*Command).Execute(...)
> 	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
> main.main()
> 	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/main.go:32 +0x105
> ```

I think we should check for nil for "clientConfig.service" also, I fixed at v0.0.6, can you please retry with v0.0.6?

**Comment by ahmetb** [comment]
Still doesn't work, different error. 

It seems you've written the code to not nil-check in multiple places, I'm guessing there are many more.
 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1383764]

goroutine 1 [running]:
github.com/Trendyol/kubectl-view-webhook/pkg/printer.(*Printer).Print(0xc000597cf0, 0xc0003a3440)
	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/pkg/printer/printer.go:142 +0x644
github.com/Trendyol/kubectl-view-webhook/cmd.(*ViewWebhookOptions).Run(0xc00044f6c0, 0xc00042d340, 0x2cf6778)
	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/cmd/root.go:139 +0xe5
github.com/Trendyol/kubectl-view-webhook/cmd.NewCmdViewWebhook.func1(0xc00042d340, 0x2cf6778, 0x0, 0x0, 0x0, 0x0)
	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/cmd/root.go:70 +0x78
github.com/spf13/cobra.(*Command).execute(0xc00042d340, 0xc00003a1d0, 0x0, 0x0, 0xc00042d340, 0xc00003a1d0)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0xc00042d340, 0xc00000e010, 0x23d4c00, 0xc00000e018)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
main.main()
	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/main.go:32 +0x105
```

We expect a decent experience from plugins.  It might make sense to adopt a more defensive coding style if you don't know what fails when. If this trend sustains, we're likely to remove the plugin from the index.

**Comment by developer-guy** [comment]
> Still doesn't work, different error.
> 
> It seems you've written the code to not nil-check in multiple places, I'm guessing there are many more.
> 
> ```
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1383764]
> 
> goroutine 1 [running]:
> github.com/Trendyol/kubectl-view-webhook/pkg/printer.(*Printer).Print(0xc000597cf0, 0xc0003a3440)
> 	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/pkg/printer/printer.go:142 +0x644
> github.com/Trendyol/kubectl-view-webhook/cmd.(*ViewWebhookOptions).Run(0xc00044f6c0, 0xc00042d340, 0x2cf6778)
> 	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/cmd/root.go:139 +0xe5
> github.com/Trendyol/kubectl-view-webhook/cmd.NewCmdViewWebhook.func1(0xc00042d340, 0x2cf6778, 0x0, 0x0, 0x0, 0x0)
> 	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/cmd/root.go:70 +0x78
> github.com/spf13/cobra.(*Command).execute(0xc00042d340, 0xc00003a1d0, 0x0, 0x0, 0xc00042d340, 0xc00003a1d0)
> 	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x47c
> github.com/spf13/cobra.(*Command).ExecuteC(0xc00042d340, 0xc00000e010, 0x23d4c00, 0xc00000e018)
> 	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x375
> github.com/spf13/cobra.(*Command).Execute(...)
> 	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
> main.main()
> 	/home/runner/work/kubectl-view-webhook/kubectl-view-webhook/main.go:32 +0x105
> ```
> 
> We expect a decent experience from plugins. It might make sense to adopt a more defensive coding style if you don't know what fails when. If this trend sustains, we're likely to remove the plugin from the index.

You're totally right, sorry about the bad experience but it's hard to guess which files whether nil or not, in v0.0.7 seems everything works fine, also in this plugin we should need to add our unit&integration test both to figure out this type of problems.

**Comment by ahmetb** [comment]
Nice. Worked now.

The output wasn't the most useful. OPERATION column seemed repetitive. But you can address those later.

Also, please consider setting up [Krew release automation](https://krew.sigs.k8s.io/docs/developer-guide/release/automating-updates/) which helps you skip manually making updates to your Krew manifests on each new version and send a pull request. It’s a GitHub workflow bot that can run every time you push a new tag. Those PRs are automatically approved.

image

**Comment by developer-guy** [comment]
> Nice. Worked now.
> 
> The output wasn't the most useful. OPERATION column seemed repetitive. But you can address those later.
> 
> Also, please consider setting up [Krew release automation](https://krew.sigs.k8s.io/docs/developer-guide/release/automating-updates/) which helps you skip manually making updates to your Krew manifests on each new version and send a pull request. It’s a GitHub workflow bot that can run every time you push a new tag. Those PRs are automatically approved.
> 
> image

Thanks for your support and quick responses, you're the best 🙏 I will fix it too 👍

**Pull request merged**